swimlane/core/resources/app_revision.py

```diff
@@ -0,0 +1,46 @@
+import pendulum
+
+from swimlane.core.resources.app import App
+from swimlane.core.resources.base import APIResource
+from swimlane.core.resources.usergroup import UserGroup
+
+
+class AppRevision(APIResource):
```

Comment:
Like the approach and expansion here, only concern is that the `AppRevision`/`App` + `RecordRevision`/`Record` are effectively the same things, but have no common inheritance. Seems like quite a bit of copy-pasta in these Revision classes.

Seems like if we really want caching support for these revisions, these classes could just extend the base classes, or better yet find a way to make those base classes respect their revision info if possible.

Reply:
Good points! I was thinking about adding a base class for Revisions but it slipped my mind. Do you think adding a revision base class would eliminate the duplication here?

It sounds like you may also be suggesting that Revisions should inherit from App / Record. On the backend models, a Revision contains a App/Record rather than inheriting from it so I think that makes sense here as well. At the moment I don't see how extending App/Record on the Revision classes would simplify anything but I'm open to it.

In terms of making base classes respect their revision info, could you expand on that idea? I'm not quite sure what you mean there.

Reply:
Yep, I was suggesting ideally looking for option that keeps a single `App` + `Record` class, and have the revision adapters you added use those instead, modifying those two classes to look at the raw data to see if they are revisions or the current active version. Would have to dig in a bit to see where the shortest option for that is, just seemed like with as much as was copied in, they didn't seem to diverge all that much at first glance.

Reply:
Just committed a base Revision class which should take care of duplication between RecordRevision and AppRevision. What do you see that is duplicated between App and AppRevision, or Record and RecordRevision? For App and AppRevision I see no shared attributes or methods. For Record and RecordRevision, they both have an `app` attribute but one is private. Not seeing how basing AppRevision or RecordRevision on App/Record would simplify things yet?